heartbeat/service.go

```diff
@@ -0,0 +1,175 @@
+// Copyright (C) 2019-2024 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package heartbeat
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger/eval"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// Service emits keep-alive heartbeats for accts that are in danger of
+// suspension.
+type Service struct {
+	// addresses that should be monitored for suspension
+	accts participants
+	// current status and balances
+	ledger ledger
+	// where to send the heartbeats
+	bcast txnBroadcaster
+
+	// infrastructure
+	ctx      context.Context
+	shutdown context.CancelFunc
+	wg       sync.WaitGroup
+	log      logging.Logger
+}
+
+// NewService creates a heartbeat service. It will need to know which accounts
+// to emit heartbeats for, and how to create the heartbeats.
+func NewService(accts participants, ledger ledger, bcast txnBroadcaster, log logging.Logger) *Service {
+	return &Service{
+		accts:  accts,
```

Comment:
I think it will handle that. `accts` is a `participants`, not just a slice of addresses or similar.  I assume the `AccountManager`'s `Keys()` method updates if participants are added.

```
// partipants captures the aspects of the AccountManager that are used by this
// package. Service must be able to find out which accounts to monitor and have
// access to their part keys to construct heartbeats.
type participants interface {
	Keys(rnd basics.Round) []account.ParticipationRecordForRound
}
```

I'm trying to follow some of the other services by creating these interface types that specify only what the service needs, making mocking simpler. I'm not entirely sure it's worthwhile.

Reply:
Indeed, works as intended